src/main/java/org/logstash/beats/BeatsParser.java

```diff
@@ -0,0 +1,230 @@
+
+                batch.setBatchSize((int) in.readUnsignedInt());
+
+                // This is unlikely to happen but I have no way to known when a frame is
+                // actually completely done other than checking the windows and the sequence number,
+                // If the FSM read a new window and I have still
+                // events buffered I should send the current batch down to the next handler.
+                if(!batch.isEmpty()) {
+                    logger.warn("New window size received but the current batch was not complete, sending the current batch");
+                    out.add(batch);
+                    batchComplete();
+                }
+
+                transitionToReadHeader();
+                break;
+            }
+            case READ_DATA_FIELDS: {
+                // Lumberjack version 1 protocol, which use the Key:Value format.
+                logger.debug("Running: READ_DATA_FIELDS");
+                sequence = (int) in.readUnsignedInt();
+                int fieldsCount = (int) in.readUnsignedInt();
+                int count = 0;
+
+                Map dataMap = new HashMap<String, String>();
+
+                while(count < fieldsCount) {
+                    int fieldLength = (int) in.readUnsignedInt();
+                    String field = in.readBytes(fieldLength).toString(Charset.forName("UTF8"));
+
+                    int dataLength = (int) in.readUnsignedInt();
+                    String data = in.readBytes(dataLength).toString(Charset.forName("UTF8"));
+
+                    dataMap.put(field, data);
+
+                    count++;
+                }
+
+                Message message = new Message(sequence, dataMap);
+                batch.addMessage(message);
+
+                if(batch.complete()) {
+                    out.add(batch);
+                    batchComplete();
+                }
+
+                transitionToReadHeader();
+
+                break;
+            }
+            case READ_JSON_HEADER: {
+                logger.debug("Running: READ_JSON_HEADER");
+
+                sequence = (int) in.readUnsignedInt();
+                int jsonPayloadSize = (int) in.readUnsignedInt();
+
+                transition(States.READ_JSON, jsonPayloadSize);
+                break;
+            }
+            case READ_COMPRESSED_FRAME_HEADER: {
+                logger.debug("Running: READ_COMPRESSED_FRAME_HEADER");
+
+                transition(States.READ_COMPRESSED_FRAME, in.readUnsignedInt());
+                break;
+            }
+
+            case READ_COMPRESSED_FRAME: {
+                logger.debug("Running: READ_COMPRESSED_FRAME");
+
+
+                byte[] bytes = new byte[(int) requiredBytes];
+                in.readBytes(bytes);
+
+                InputStream inflater = new InflaterInputStream(new ByteArrayInputStream(bytes));
+                ByteArrayOutputStream decompressed = new ByteArrayOutputStream();
```

Comment:
Another thing that comes to mind here and might be worth testing to prevent copying byte arrays around too much:

use a `InflaterOutputStream` and a `ByteBufOutputStream` and call `in.readBytes(OutputStream os, int length)`